### PR TITLE
fix/rec-iql and sac timestep logging

### PR DESCRIPTION
**Pull request by JemmaLDaniel**
## What?
1. Extend the experiment's main loop by one to run up to and inclusive of `total_timesteps`.
2. Log `steps_per_second` as actor, evaluator and absolute logging events, instead of miscellaneous logging events.
## Why?
Both of these fixes align IQL and SAC with the JSON logging of Mava's PPO systems, enabling aggregated plotting over all systems. Currently, aggregate plotting between PPO systems and IQL/SAC systems will error with non-matching steps and metrics.

### CODE REVIEW

File: mava/systems/q_learning/rec_iql.py [hunk 1]
```diff
@@ -564,33 +564,36 @@ def run_experiment(cfg: DictConfig) -> float:
...
         # learn steps is not included in the loop counter.
-        learn_steps = anakin_steps * cfg.system.epochs
-        sps = (t + learn_steps) / (time.time() - start_time)
+        elapsed_time = time.time() - start_time
         eps = jax.numpy.maximum(
             cfg.system.eps_min, 1 - (t / cfg.system.eps_decay) * (1 - cfg.system.eps_min)
         )
-
         final_metrics, ep_completed = episode_metrics.get_final_step_metrics(metrics)
+        final_metrics["steps_per_second"] = steps_per_rollout / elapsed_time
```
Comment (sash-a) [inline mava/systems/q_learning/rec_iql.py:586]: So this is why I originally combined the two steps per second values. It's not fair to say that we do `steps_per_rollout / elapsed_time` because we aren't only doing `steps_per_rollout` actor steps here, we are also doing `anakin_steps * cfg.system.epochs` gradient steps here. 

So it's not really a true representation of either the learner steps per second or the actor steps per second.

Could you maybe put this in a comment above or make a new issue to fix this in all the all the systems :pray:

File: mava/systems/q_learning/rec_iql.py [hunk 2]
```diff
@@ -599,7 +602,10 @@ def run_experiment(cfg: DictConfig) -> float:
         jax.block_until_ready(eval_output)
 
         # Log:
+        elapsed_time = time.time() - start_time
         episode_return = jnp.mean(eval_output.episode_metrics["episode_return"])
+        steps_per_eval = int(jnp.sum(eval_output.episode_metrics["episode_length"]))
```
Comment (sash-a) [inline mava/systems/q_learning/rec_iql.py:607]: No need to make this an int as you're diving with it on the next line. Unless it's to get rid of the `Array` around it? In which case you can leave it

```suggestion
        steps_per_eval = jnp.sum(eval_output.episode_metrics["episode_length"])
```

File: mava/systems/q_learning/rec_iql.py [hunk 3]
```diff
@@ -621,11 +627,17 @@ def run_experiment(cfg: DictConfig) -> float:
...
+        start_time = time.time()
+
         eval_keys = jax.random.split(key, cfg.arch.n_devices)
 
         eval_output = absolute_metric_evaluator(best_params, eval_keys)
         jax.block_until_ready(eval_output)
 
+        elapsed_time = time.time() - start_time
+
+        steps_per_eval = int(jnp.sum(eval_output.episode_metrics["episode_length"]))
```
Comment (sash-a) [inline mava/systems/q_learning/rec_iql.py:639]: Same here
```suggestion
        steps_per_eval = jnp.sum(eval_output.episode_metrics["episode_length"])
```

File: mava/systems/sac/ff_isac.py [hunk 4]
```diff
@@ -517,47 +517,56 @@ def run_experiment(cfg: DictConfig) -> float:
...
         # Evaluate:
         key, eval_key = jax.random.split(key)
         eval_keys = jax.random.split(eval_key, cfg.arch.n_devices)
         eval_output = evaluator(unreplicate_batch_dim(learner_state.params.actor), eval_keys)
         jax.block_until_ready(eval_output)
 
         # Log:
+        elapsed_time = time.time() - start_time
         episode_return = jnp.mean(eval_output.episode_metrics["episode_return"])
+        steps_per_eval = int(jnp.sum(eval_output.episode_metrics["episode_length"]))
```
Comment (sash-a) [inline mava/systems/sac/ff_isac.py:568]:
```suggestion
        steps_per_eval = jnp.sum(eval_output.episode_metrics["episode_length"])
```

File: mava/systems/sac/ff_isac.py [hunk 5]
```diff
@@ -580,11 +589,17 @@ def run_experiment(cfg: DictConfig) -> float:
...
+        start_time = time.time()
+
         eval_keys = jax.random.split(key, cfg.arch.n_devices)
 
         eval_output = absolute_metric_evaluator(best_params, eval_keys)
         jax.block_until_ready(eval_output)
 
+        elapsed_time = time.time() - start_time
+
+        steps_per_eval = int(jnp.sum(eval_output.episode_metrics["episode_length"]))
```
Comment (sash-a) [inline mava/systems/sac/ff_isac.py:601]:
```suggestion
        steps_per_eval = jnp.sum(eval_output.episode_metrics["episode_length"])
```

File: mava/systems/sac/ff_isac.py [hunk 6]
```diff
@@ -517,47 +517,56 @@ def run_experiment(cfg: DictConfig) -> float:
...
 
         # Log:
         # Add learn steps here because anakin steps per second is learn + act steps
         # But we also want to make sure we're counting env steps correctly so
         # learn steps is not included in the loop counter.
-        learn_steps = anakin_steps * cfg.system.epochs
-        sps = (t + learn_steps) / (time.time() - start_time)
+        elapsed_time = time.time() - start_time
         final_metrics, ep_completed = episode_metrics.get_final_step_metrics(metrics)
+        final_metrics["steps_per_second"] = steps_per_rollout / elapsed_time
```
Comment (sash-a) [inline mava/systems/sac/ff_isac.py:548]: Same here, I think just make an issue to fix this

File: mava/systems/sac/ff_masac.py [hunk 7]
```diff
@@ -538,44 +538,54 @@ def run_experiment(cfg: DictConfig) -> float:
...
         # Evaluate:
         key, eval_key = jax.random.split(key)
         eval_keys = jax.random.split(eval_key, cfg.arch.n_devices)
         eval_output = evaluator(unreplicate_batch_dim(learner_state.params.actor), eval_keys)
         jax.block_until_ready(eval_output)
 
         # Log:
+        elapsed_time = time.time() - start_time
         episode_return = jnp.mean(eval_output.episode_metrics["episode_return"])
+        steps_per_eval = int(jnp.sum(eval_output.episode_metrics["episode_length"]))
```
Comment (sash-a) [inline mava/systems/sac/ff_masac.py:587]:
```suggestion
        steps_per_eval = jnp.sum(eval_output.episode_metrics["episode_length"])
```

File: mava/systems/sac/ff_masac.py [hunk 8]
```diff
@@ -598,11 +608,17 @@ def run_experiment(cfg: DictConfig) -> float:
...
+        start_time = time.time()
+
         eval_keys = jax.random.split(key, cfg.arch.n_devices)
 
         eval_output = absolute_metric_evaluator(best_params, eval_keys)
         jax.block_until_ready(eval_output)
 
+        elapsed_time = time.time() - start_time
+
+        steps_per_eval = int(jnp.sum(eval_output.episode_metrics["episode_length"]))
```
Comment (sash-a) [inline mava/systems/sac/ff_masac.py:620]:
```suggestion
        steps_per_eval = jnp.sum(eval_output.episode_metrics["episode_length"])
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Looking good, just 2 minor comments here :smile:

File: mava/systems/q_learning/rec_iql.py (hunk 1 above)
Comment (JemmaLDaniel) [reply mava/systems/q_learning/rec_iql.py:586]: For sure, I will make an issue!

File: mava/systems/q_learning/rec_iql.py (hunk 2 above)
Comment (JemmaLDaniel) [reply mava/systems/q_learning/rec_iql.py:607]: I followed the PPO logging method as closely as possible, which does use the `int()` here to get rid of the `Array` around `steps_per_second`.

**Review by sash-a — APPROVED** [review]
Thanks @JemmaLDaniel!

Don't forget the issue when me merge :pray:

**Review by OmaymaMahjoub — APPROVED** [review]
Thanks @JemmaLDaniel :pray:
